File: wallet-kit/src/main/java/bitcoin/walllet/kit/struct/TxOut.java
package bitcoin.walllet.kit.struct;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.io.IOException;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.common.serializer.SatoshiSerializer;

public class TxOut {

    /**
     * int64, Transaction Value
     */
    @JsonSerialize(using = SatoshiSerializer.class)
    public long value;

    /**
     * uchar[], Usually contains the public key as a Bitcoin script setting up
     * conditions to claim this output.
     */
    public byte[] pk_script;

    public TxOut(BitcoinInput input) throws IOException {
        this.value = input.readLong();
        long scriptLength = input.readVarInt();
        this.pk_script = input.readBytes((int) scriptLength);
    }

    public byte[] toByteArray() {
        return new BitcoinOutput().writeLong(value).writeVarInt(this.pk_script.length).write(pk_script).toByteArray();
    }
}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/models/TxOut.kt
package bitcoin.wallet.kit.models

import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import io.realm.RealmObject
import java.io.IOException

/**
 * Transaction output
 *
 *  Size        Field               Description
 *  ====        =====               ===========
 *  8 bytes     TxOutValue          Value expressed in Satoshis (0.00000001 BTC)
 *  VarInt      TxOutScriptLength   Script length
 *  Variable    TxOutScript         Script
 */
open class TxOut : RealmObject {

    /**
     * int64, Transaction Value
     */
    // @JsonSerialize(using = SatoshiSerializer::class)
    var value: Long = 0

    /**
     * uchar[], Usually contains the public key as a Bitcoin script setting up
     * conditions to claim this output.
     */
    var pkScript: ByteArray = byteArrayOf()

    constructor()

    @Throws(IOException::class)
    constructor(input: BitcoinInput) {
        value = input.readLong()
        val scriptLength = input.readVarInt()
        pkScript = input.readBytes(scriptLength.toInt())
    }

    fun toByteArray(): ByteArray {
        return BitcoinOutput()
                .writeLong(value)
                .writeVarInt(pkScript.size.toLong())
                .write(pkScript)
                .toByteArray()
    }
}
